### PR TITLE
Clarify behavior of groupby default=None.

### LINKED ISSUE #744: No more KeyError after changing groupbydoc to groupby
### Description

When I was adding tests for fixing the [deprecation of groupbydoc](https://github.com/glotzerlab/signac/issues/328), I noticed that there is a [test that looks for a KeyError](https://github.com/glotzerlab/signac/blob/95157b43eddde667c99b941052cc23c1304fddd9/tests/test_project.py#L1073) that doesn't raise one with the new doc query namespace. This leads me to wonder if the sp query namespace should also raise a KeyError

Additionally, should an empty `project.groupby()` raise an error?

### To reproduce

```python
import signac

project = signac.init_project("project")

for i in range(3):
    job = project.open_job(
        dict(
            a=1,
            b=i
        )
    ).init()
    job.doc = dict(
        a=1,
        b=2,
        c=i
    )


# This raise a KeyError
for k,g in project.groupbydoc('ff'):
    print(k)


# These all don't raise errors
for k,g in project.groupby('doc.ff'):
    print(k, list(g))

for k, g in project.groupby('sp.ff'):
    print(k, list(g))

for k,g in project.groupby('ff'):
    print(k, list(g))
    
for k,g in project.groupbydoc():
    print(k, list(g))

for k,g in project.groupby():
    print(k, list(g))
```

### PR DESCRIPTION
**Pull request by bdice**
## Description
Resolves #744.

## Motivation and Context
Clarifies docs. No changelog required.

## Checklist:
- [x] I am familiar with the [Contributing Guidelines](https://github.com/glotzerlab/signac/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [Contributor Agreement](https://github.com/glotzerlab/signac/blob/master/ContributorAgreement.md).
- [x] My name is on the [list of contributors](https://github.com/glotzerlab/signac/blob/master/contributors.yaml).
- [x] The changes introduced by this pull request are covered by existing or newly introduced tests.
- [x] The [package documentation](https://github.com/glotzerlab/signac/tree/master/doc) and [framework documentation](https://docs.signac.io/) in [signac-docs](https://github.com/glotzerlab/signac-docs) are up to date with these changes.
- [x] I have updated the [changelog](https://github.com/glotzerlab/signac/blob/master/changelog.txt) and added any related issue and pull request numbers for future reference.

### CODE REVIEW

**Review by vyasr — APPROVED** [review]
This LGTM but I'd like to get @cbkerr's opinion before merging just in case.

**Review by iblanco11981870 — APPROVED** [review]
looks good to me

**Review by cbkerr — APPROVED** [review]
I don't have time to make sure the description matches the code, but I corrected some other aspects of the `groupby` doc string.

File: signac/contrib/project.py [hunk 1]
```diff
@@ -2694,17 +2704,28 @@ def groupby(self, key=None, default=None):
...
+            A default value to be used when a given key is not
             present. The value must be sortable and is only used if not None
             (Default value = None).
 
+        Returns
+        -------
+        key : str
+            Grouped key.
+        group : iterable of Jobs
+            Iterable of `Job` instances matching this group key.
```
Comment (vyasr) [inline signac/contrib/project.py:2727]:
```suggestion
        Tuple[str, Iterable[Job]]
            A pair whose first element is the grouped key and whose second element
            is an iterable of `Job` instances matching the key.
```

**Comment by bdice** [comment]
@vyasr @cbkerr The changes made after my commits (changes to the "Returns") are incorrect. `itertools.groupby` returns a generator. The generated values have the structure of `Tuple[str, Iterable[Job]]`. I think that aligning with numpydoc would suggest a pattern like this: https://numpydoc.readthedocs.io/en/latest/format.html#yields

```
Yields
------
key :
    Grouped key.
group : iterable of Jobs
    Iterable of `Job` instances matching this group key.
```

Also, we do _not_ know the type of `key` (though it is typed as `str` in the changes made in this PR) because a provided callable can return any type that can be sorted and equality-compared, e.g. `bool` or `int`.

I can file a second PR to fix this, but it might have been nice to let this one sit for a little longer instead of making substantial changes and then merging so quickly...

**Comment by vyasr** [comment]
Ack sorry I wanted to get everything merged ASAP to let the release process kicked off and I didn't expect docs changes to cause problems.

**Pull request merged**